quil-rs/src/expression.rs

```diff
@@ -253,6 +253,537 @@ fn is_small(x: f64) -> bool {
+                    list @ Sexp::List(..) => {
+                        Err(SimplificationError::UnexpectedListSexp(list.clone()))
+                    }
+                    Sexp::String(op) => {
+                        let arg_ids: Vec<Id> = list[1..]
+                            .iter()
+                            .map(|s| parse_sexp_into(s, expr))
+                            .collect::<Result<_, _>>()?;
+                        let node = Expr::from_op(op, arg_ids)
+                            .map_err(|e| SimplificationError::UnknownOp(format!("{e:#?}")))?;
+                        Ok(expr.add(node))
+                    }
+                },
+            }
+        }
+        let mut expr = RecExpr::default();
+        parse_sexp_into(sexp, &mut expr)?;
+        Ok(expr)
+    }
+
+    /// Our analysis will perform arithmetic simplification (largely, constant folding) on our
+    /// language.
+    #[derive(Default)]
+    struct Arithmetic;
+    type EGraph = egg::EGraph<Expr, Arithmetic>;
+
+    /// Our analysis will perform constant folding on our language.
+    impl egg::Analysis<Expr> for Arithmetic {
+        /// Constant values
+        type Data = Option<C>;
+
+        /// Pull the (possible) [`Self::Data`] from the given expression.
+        fn make(egraph: &EGraph, enode: &Expr) -> Self::Data {
+            let x = |id: &Id| egraph[*id].data.as_ref();
+            match enode {
+                Expr::Pi => Some(C::PI),
+                Expr::Number(c) => Some(*c),
+                Expr::Cis(id) => Some(x(id)?.cis()),
+                Expr::Cos(id) => Some(x(id)?.cos()),
+                Expr::Exp(id) => Some(x(id)?.exp()),
+                Expr::Sin(id) => Some(x(id)?.sin()),
+                Expr::Sqrt(id) => Some(x(id)?.sqrt()),
+                Expr::Pos(id) => Some(*x(id)?),
+                Expr::Neg(id) => Some(-*x(id)?),
+                Expr::Pow([base, power]) => Some(x(base)?.pow(*x(power)?)),
+                Expr::Mul([left, right]) => Some(*x(left)? * *x(right)?),
+                Expr::Div([left, right]) => Some(*x(left)? / *x(right)?),
+                Expr::Add([left, right]) => Some(*x(left)? + *x(right)?),
+                Expr::Sub([left, right]) => Some(*x(left)? - *x(right)?),
+                _ => None,
+            }
+        }
+
+        /// Merge two pieces of data with the same value.
+        fn merge(&mut self, to: &mut Self::Data, from: Self::Data) -> egg::DidMerge {
+            egg::merge_option(to, from, |a, b| {
+                assert!(a.close(b), "Merged non-equal constants");
+                egg::DidMerge(false, false)
+            })
+        }
+
+        /// Update the graph to equate and simplify constant values.
+        fn modify(egraph: &mut EGraph, id: Id) {
+            if let Some(c) = egraph[id].data {
+                let value = match c {
+                    _ if c.close(C::PI) => Expr::Pi,
+                    _ => Expr::Number(c),
+                };
```

Comment:
This should be the same logic, unless there's something I'm not understanding here

```suggestion
                let value = if c.close(C::PI) {
                    Expr::Pi
                } else {
                    Expr::Number(c)
                };
```